### PR TITLE
[#406] LogStructured

### PR DESCRIPTION
**Pull request by CodiePP**
description
-----------

- [ ] `LogObject` now has constructor `LogStructured` to capture arbitrary structured log message

checklist
---------

- [x] compiles (`cabal new-clean; cabal new-build`)
- [x] tests run successfully (`cabal new-test`)
- [x] documentation added and created (`cd docs; nix-shell --run make`)
- [x] link to an issue
- [x] link to an epic
- [x] add estimate points
- [x] add milestone (the same as the linked issue)

### CODE REVIEW

File: iohk-monitoring/src/Cardano/BM/Data/LogItem.lhs [hunk 1]
```diff
@@ -60,8 +66,20 @@ data LogObject a = LogObject
                      { loName    :: LoggerName
                      , loMeta    :: !LOMeta
                      , loContent :: (LOContent a)
-                     }
-                     deriving (Generic, Show, ToJSON, FromJSON)
+                     } deriving (Show, Eq)
+
+instance ToJSON a => ToJSON (LogObject a) where
```
[the remark this answers is not in the document]
Comment (ksaric) [reply iohk-monitoring/src/Cardano/BM/Data/LogItem.lhs]: So all the content of a `LogObject` must be serializable to JSON?

[the remark this answers is not in the document]
Comment (CodiePP) [reply iohk-monitoring/src/Cardano/BM/Data/LogItem.lhs]: yes

File: iohk-monitoring/src/Cardano/BM/Data/LogItem.lhs [hunk 2]
```diff
@@ -60,8 +66,20 @@ data LogObject a = LogObject
...
+    toJSON (LogObject _loname _lometa _locontent) =
+        object [ "loname"    .= _loname
+               , "lometa"    .= _lometa
+               , "locontent" .= _locontent
+               ]
+instance (FromJSON a) => FromJSON (LogObject a) where
+    parseJSON j = withObject "LogObject"
```
[the remark this answers is not in the document]
Comment (CodiePP) [reply iohk-monitoring/src/Cardano/BM/Data/LogItem.lhs]: yes, better.

File: iohk-monitoring/src/Cardano/BM/Data/LogItem.lhs [hunk 3]
```diff
@@ -173,7 +305,14 @@ data CommandValue = DumpBufferedTo BackendKind
 data PrivacyAnnotation =
       Confidential -- confidential information - handle with care
     | Public       -- indifferent - can be public.
-    deriving (Show, Generic, ToJSON, FromJSON)
+    deriving (Show, Eq)
+
+instance FromJSON PrivacyAnnotation where
+    parseJSON v = withText "PrivacyAnnotation"
```
[the remark this answers is not in the document]
Comment (CodiePP) [reply iohk-monitoring/src/Cardano/BM/Data/LogItem.lhs]: 👍

**Pull request merged**